### ISSUE TITLE
IsBitwiseMoveable compilation failed for c++11

### ISSUE THREAD

**Issue by lboquillon** [bloomberg/bde#202]
Hi guys,

I'm trying to compile Bloomberg BSL group with gcc version 5.3.0 and -std=c++11 flag and I got this error[1].

If I Understand Correctly I should template-specialize the type, but since this is a native Bloomberg BDE type (bslalg::HashTableAnchor), I think this should be natively supported.

Could you please take a look, and if the triage results that this should be fixed, could you also provide an ETA?

We need to get some planning clue in order to decide when to implement the stuff that will use this code.

[1]=

> build/bde/groups/bsl/bslmf/bslmf_isbitwisemoveable.h: In instantiation of 'struct BloombergLP::bslmf::IsBitwiseMoveable_Imp<BloombergLP::bslalg::HashTableAnchor, false>':
> build/bde/groups/bsl/bslmf/bslmf_isbitwisemoveable.h:677:8:   required from 'struct BloombergLP::bslmf::IsBitwiseMoveable<BloombergLP::bslalg::HashTableAnchor>'
> build/bde/groups/bsl/bslalg/bslalg_scalarprimitives.h:2223:17:   required from 'static void BloombergLP::bslalg::ScalarPrimitives::swap(LHS_TYPE&, RHS_TYPE&) [with LHS_TYPE = BloombergLP::bslalg::HashTableAnchor; RHS_TYPE = BloombergLP::bslalg::HashTableAnchor]'
> build/bde/groups/bsl/bslalg/bslalg_hashtableanchor.h:671:48:   required from here
> build/bde/groups/bsl/bslmf/bslmf_isbitwisemoveable.h:667:5: error: static assertion failed: False positive IsBitwiseMoveable for one-byte type
>      static_assert(! (value && !k_ValueWithoutOnebyteHeuristic),
>      ^
> scons: *** [build/bde/groups/bsl/bslalg/bslalg_hashtableanchor.o] Error 1

Thanks,
Leonardo Boquillón

**Comment by mversche** [comment]
gcc 5.x isn't currently in our list of supported compilers.  I expect it will eventually be added, but its not yet in our nightly builds (or necessarily easily available to test with).  

I believe @abeels tested some flavor of gcc 5.x recently, found errors, which I thought were identified and fixed, but possibly not merged.  I'll ask him to comment, though he won't see this until at least tomorrow (out sick).

**Comment by lboquillon** [comment]
@mversche thanks.

I've made it compile adding

```cpp
namespace bslmf {
template <>
struct IsBitwiseMoveable<bslalg::HashTableAnchor>
    : bsl::false_type {};
} 
```

Maybe that can hint you what the problem is.

**Comment by dgutson** [comment]
@mversche , IMHO I don't see any compiler-specific issue here but a pure C++(11) one.
Have you ever tried this with C++11 regardless of the compiler version? (of course, the implementation degree depends on the compiler version, but I don't think this is a matter of unsopported features, but mis-use of already supported features).

**Comment by AlisdairM** [comment]
Yes, we regularly test with gcc 4.9 in C++11 mode, the problem is that the trivial trait we test in a static_assert gets intrinsic support in the gcc 5.0 library.  For some in-progress work, we need a reliable implementation of our bitwise movable trait for empty types (like std library functors) and that is really hard to test reliably while supporting C++03, due to the possibility of unions - which we cannot isolate without a compiler intrinsic in the 'is_union' trait, and requiring C++11 again.  We make an assumption that any type smaller than a pointer cannot have an internal reference, and so satisfies the constraints for our bitwise movable trait.  This will not hold for types that store their address or manage state in some external registry, but is a good answer for the overwhelmingly common case.  We have a static_assert that fires in a C++11 build for non-empty types that are also non-trivial, but would deduce as bitwise-movable according to our formula for C++03.  The common case is to then specialize the trait to 'true', as most cases we have examined turn out to still satisfy our trait.  However, this is the compiler finding code that should be audited, and explicitly rather than implicitly flagged for the trait, when the toolchain has the intrinsics available to test our assumptions.

**Comment by mversche** [comment]
It took me a bit to parse that :) Here's my TLDR to summarize:

This works in gcc-4.9 (C++11), but fails in gcc-5.x (C++11) because gcc 5.x has support for additional type traits.  More specifically, the `IsBitwiseMoveable` trait acquires a default value based on some C++11 traits that we must emulate for C++03 compilers; this failing static assertion is indicating a place where the derived `IsBitwiseMoveable` property is different when determined using the emulation than with the actual compiler traits. 

A quick fix would be what @lboquillon suggested, though I believe HashtableAnchor should be specialized to `true` for the `IsBitwiseMoveableTrait`.  I believe @abeels had such a change, and can confirm tomorrow.  The long-term solution is some changes to the `type_traits` implementations/emulations catching some edge cases.  I believe that is work @AlisdairM  has in progress.

**Comment by abeels** [comment]
Yes, we already have a fix for this available.  We should be able to push it out on Friday, with a raft of other updates.

**Comment by abeels** [comment]
The fix for this issue is causing breakage on some of our platforms, so it didn't make it into today's GitHub update.  @AlisdairM is going to be working on related issues next week, and we should be able to roll out the fix once those issues are resolved.

**Comment by abeels** [comment]
Fixed by d66b60bc05cab18f6dae332878768c38078711dc
